avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -587,97 +569,114 @@ private void processArray(JVar arraySchemaVar, final String name, final Schema a
+    final JVar arrayVar = action.getShouldRead() ? declareValueVar(name, arraySchema, parentBody, true, false, true) : null;
     /**
      * Special optimization for float array by leveraging {@link ByteBufferBackedPrimitiveFloatList}.
+     *
+     * TODO: Handle other primitive element types here.
      */
     if (action.getShouldRead() && arraySchema.getElementType().getType().equals(Schema.Type.FLOAT)) {
       JClass primitiveFloatList = codeModel.ref(ByteBufferBackedPrimitiveFloatList.class);
       JExpression readPrimitiveFloatArrayInvocation = primitiveFloatList.staticInvoke("readPrimitiveFloatArray").
           arg(reuseSupplier.get()).arg(JExpr.direct(DECODER));
       JExpression castedResult =
-          JExpr.cast(codeModel.ref(List.class).narrow(Float.class), readPrimitiveFloatArrayInvocation);
+          JExpr.cast(codeModel.ref(PrimitiveFloatList.class), readPrimitiveFloatArrayInvocation);
 
       parentBody.assign(arrayVar, castedResult);
       putArrayIntoParent.accept(parentBody, arrayVar);
       return;
     }
 
-    JClass arrayClass = schemaAssistant.classFromSchema(arraySchema, false);
+    JClass arrayClass = schemaAssistant.classFromSchema(arraySchema, false, false, true);
+    JClass abstractErasedArrayClass = schemaAssistant.classFromSchema(arraySchema, true, false, true).erasure();
 
     JVar chunkLen =
         parentBody.decl(codeModel.LONG, getUniqueName("chunkLen"), JExpr.direct(DECODER + ".readArrayStart()"));
 
-    JConditional conditional = parentBody._if(chunkLen.gt(JExpr.lit(0)));
-    JBlock ifBlockForChunkLenCheck = conditional._then();
-    if (action.getShouldRead()) {
-      JVar reuse = declareValueVar(name + "Reuse", arraySchema, ifBlockForChunkLenCheck);
+    final FieldAction finalAction = action;
+    JInvocation newArrayExp = JExpr._new(arrayClass);
+    if (useGenericTypes) {
+      newArrayExp = newArrayExp.arg(JExpr.cast(codeModel.INT, chunkLen));
+      if (!SchemaAssistant.isPrimitive(arraySchema.getElementType())) {
+        /**
+         * N.B.: The ColdPrimitiveXList implementations do not take the schema as a constructor param,
+         * but the {@link org.apache.avro.generic.GenericData.Array} does.
+         */
+        newArrayExp = newArrayExp.arg(getSchemaExpr(arraySchema));
+      }
+    }
+    JInvocation finalNewArrayExp = newArrayExp;
+
+    JExpression chunkLengthGreaterThanZero = chunkLen.gt(JExpr.lit(0));
+
+    ifCodeGen(parentBody, chunkLengthGreaterThanZero, then1 -> {
+      if (finalAction.getShouldRead()) {
+        JExpression conditionForReuse = reuseSupplier.get()
+            /** The reuse param must not be null */
+            .ne(JExpr.direct("null"))
+            /**
+             * And it must be an instance of the right type.
+             * N.B.: Need to use the erasure because instanceof does not support generic types
+             */
+            .cand(reuseSupplier.get()._instanceof(abstractErasedArrayClass));
```

Comment:
Nitpick: looks like we don't need to validate "null" here since `_instanceof(abstractErasedArrayClass)` also covers null case?

Reply:
Does it? I didn't know that.